### PR TITLE
the refactor of collecting accounts' storage usage info 

### PR DESCRIPTION
**Pull request by gouhongshen**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [x] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #11480

## What this PR does / why we need it:

design doc: https://github.com/gouhongshen/docs/blob/main/storage_usage.md

changes:

1. CN side

    1. re-implement doShowAccounts.
        
        * removed built-in functions `mo_table_size` and `mo_table_rows` in SQL `getTableStatsFormatV2`.
        * changed the view schema when `show accounts`, removed the `rows` column
        * get table size info by quering TN. TN sends ckp location back and then CN loads the ckp and decodes it. the query  go through the `txnOperator.Debug` pipeline.
        
2. TN side
    
    1. added a new batch into checkpoint.
        
        * schema: `[account_id, database_id, table_id, table_total_size]`
        * storing the full datasets in each checkpoint
   

```
sql client
         \
          ===> show accounts --1--> get accounts info from system table [account_id, db count, tbl count, account create time ....]
         /          |
cron task           2 <--> StorageUsage request --> TN --> the newest ckp location
                    |                             
                    3 --> load ckp and decode it
                    |
                    4 --> combine the two results from 1 and 3
                    |
+--------------+------------+---------------------+--------+----------------+----------+-------------+-------------+----------------+
| account_name | admin_name | created             | status | suspended_time | db_count | table_count | size_in_mb  | comment        |
+--------------+------------+---------------------+--------+----------------+----------+-------------+-------------+----------------+
| sys          | root       | 2023-10-24 03:30:30 | open   | NULL           |        7 |          86 | 1009.625265 | system account |
| acc1         | root1      | 2023-10-24 05:19:30 | open   | NULL           |        5 |          49 |    0.010583 |                |
| acc2         | root2      | 2023-10-24 05:19:43 | open   | NULL           |        5 |          49 |    0.010583 |                |
| acc3         | root3      | 2023-10-24 05:20:30 | open   | NULL           |        5 |          49 |    0.010583 |                |
| acc4         | root4      | 2023-10-26 06:05:12 | open   | NULL           |        5 |          52 |    0.010589 |                |
| acc5         | root5      | 2023-10-26 06:05:32 | open   | NULL           |        5 |          52 |    0.010583 |                |
| acc6         | root6      | 2023-10-26 06:15:21 | open   | NULL           |        5 |          52 |    0.010583 |                |
| acc7         | root7      | 2023-10-26 06:15:39 | open   | NULL           |        5 |          52 |    0.010597 |                |
+--------------+------------+---------------------+--------+----------------+----------+-------------+-------------+----------------+
```

### CODE REVIEW

File: pkg/util/export/example/main.go [hunk 1]
```diff
@@ -17,7 +17,7 @@
 // There are two type table data: one is dummyStatementTable, with primary key; another is dummyRawlogTable, without primary key.
 // Both of them are code-copied from pkg "github.com/matrixorigin/matrixone/pkg/util/trace", caused by import-cycle issue.
 //
-// If you want to run this code, you need some source data, that could copy from folder `mo-data/etl/sys/logs/...`, which generated by mo-service startup with config `etc/launch-tae-logservice`
```
Comment (xzxiong) [inline pkg/util/export/example/main.go:20]: 这个文件不用动吧？修改似乎没有必要

**Review by xzxiong — APPROVED** [review]

File: pkg/util/export/example/main.go (hunk 1 above)
Comment (gouhongshen) [reply pkg/util/export/example/main.go:20]: 这个是 IDE 改动了的，我修改回来了

File: pkg/vm/engine/tae/catalog/segment.go [hunk 2]
```diff
@@ -48,24 +49,121 @@ type SegmentEntry struct {
...
+}
+
+func (s *SegStat) GetSortKeyZonemap() index.ZM {
+	s.RLock()
+	defer s.RUnlock()
+	return s.sortKeyZonemap.Clone()
+}
+
+func (s *SegStat) SetRows(rows int) {
+	s.RLock()
```
Comment (aptend) [inline pkg/vm/engine/tae/catalog/segment.go:121]: s.Lock()

File: pkg/vm/engine/tae/catalog/segment.go [hunk 3]
```diff
@@ -48,24 +49,121 @@ type SegmentEntry struct {
...
+
+	if schema.HasSortKey() {
+		col := schema.GetSingleSortKey()
+		s.sortKeyZonemap = meta.MustGetColumn(col.SeqNum).ZoneMap()
+	}
+
+	s.loaded = true
+
+	// after loading object info, original size is still 0, we have to estimate it by experience
+	if name == motrace.RawLogTbl && s.originSize == 0 && s.rows != 0 {
```
Comment (aptend) [inline pkg/vm/engine/tae/catalog/segment.go:100]: 逻辑不对了，跳过 load 后不会再预估 object 大小

**Review by aptend — CHANGES_REQUESTED** [review]

**Review by zhangxu19830126 — APPROVED** [review]

File: pkg/vm/engine/tae/catalog/segment.go (hunk 2 above)
Comment (gouhongshen) [reply pkg/vm/engine/tae/catalog/segment.go:121]: OK，有几个都改了

File: pkg/vm/engine/tae/catalog/segment.go (hunk 3 above)
Comment (gouhongshen) [reply pkg/vm/engine/tae/catalog/segment.go:100]: 第一次 load 应该会做这件事

**Review by qingxinhome — APPROVED** [review]

**Review by m-schen — APPROVED** [review]

File: pkg/frontend/show_account.go [hunk 4]
```diff
@@ -71,19 +101,17 @@ const (
 		") as `admin_name`, " +
 		"count(distinct mt.reldatabase) as `db_count`, " +
 		"count(distinct mt.relname) as `table_count`, " +
-		"sum(mo_table_rows(mt.reldatabase,mt.relname)) as `row_count`, " +
-		"cast(sum(mo_table_size(mt.reldatabase,mt.relname))/1048576  as decimal(29,3)) as `size` " +
+		"cast(0 as double) as `size_in_mb` " +
```
Comment (daviszhen) [inline pkg/frontend/show_account.go:104]: 列名不要改，保持`size`。之前跟产品约定好的。其它列名同样不要改。

File: pkg/frontend/show_account.go [hunk 5]
```diff
@@ -60,7 +71,26 @@ const (
 	idxOfSuspendedTime = 4
 	idxOfComment       = 5
 
-	getTableStatsFormat = "select " +
+	//getTableStatsFormat = "select " +
+	//	"( select " +
```
Comment (daviszhen) [inline pkg/frontend/show_account.go:75]: 不要就直接删掉。

File: pkg/frontend/show_account.go [hunk 6]
```diff
@@ -116,7 +143,103 @@ func getSqlForAccountInfo(accountId uint64) string {
...
+	}
+
+	// two source:
+	// 	1. sql client
+	//	2. internal
+	if ses.proc.TxnOperator == nil {
+		defer func() { // without this: the transaction xxx has been committed or aborted
+			ses.proc.TxnOperator = nil
+		}()
+		if _, ses.proc.TxnOperator, err = ses.TxnCreate(); err != nil {
```
Comment (daviszhen) [inline pkg/frontend/show_account.go:167]: 取新事物。不要用txnCreate。用`func (th *TxnHandler) GetTxn()`。这么用`ses.txnHandler.GetTxn`，参照其它地方的用法。

File: pkg/frontend/show_account.go [hunk 7]
```diff
@@ -116,7 +143,103 @@ func getSqlForAccountInfo(accountId uint64) string {
...
+		if err := usage.Unmarshal(payload); err != nil {
+			return nil, err
+		}
+		return usage, nil
+	}
+
+	// two source:
+	// 	1. sql client
+	//	2. internal
+	if ses.proc.TxnOperator == nil {
```
Comment (daviszhen) [inline pkg/frontend/show_account.go:163]: 下面的163行到170行。用ses.txnHandler.GetTxn替换。参考其他地方用法。

File: pkg/objectio/metav2.go
Comment (gouhongshen) [inline pkg/objectio/metav2.go:1]: should be 26+2

File: pkg/frontend/show_account.go [hunk 8]
```diff
@@ -443,7 +537,7 @@ func mergeOutputResult(ses *Session, outputBatch *batch.Batch, rsOfMoAccount *ba
 		if err != nil {
 			return err
 		}
-		err = outputBatch.Vecs[finalIdxOfRowCount].UnionOne(bat.Vecs[idxOfRowCount], 0, mp)
+		//err = outputBatch.Vecs[finalIdxOfRowCount].UnionOne(bat.Vecs[idxOfRowCount], 0, mp)
```
Comment (daviszhen) [inline pkg/frontend/show_account.go:540]: 删掉

**Review by daviszhen — CHANGES_REQUESTED** [review]

**Review by LeftHandCold — APPROVED** [review]

File: pkg/frontend/show_account.go (hunk 4 above)
Comment (gouhongshen) [reply pkg/frontend/show_account.go:104]: 好的

File: pkg/vm/engine/tae/catalog/segment.go (hunk 3 above)
Comment (gouhongshen) [reply pkg/vm/engine/tae/catalog/segment.go]: 已更新

File: pkg/frontend/show_account.go (hunk 4 above)
Comment (gouhongshen) [reply pkg/frontend/show_account.go]: 已更新

File: pkg/frontend/show_account.go (hunk 5 above)
Comment (gouhongshen) [reply pkg/frontend/show_account.go]: 已更新

File: pkg/frontend/show_account.go (hunk 6 above)
Comment (gouhongshen) [reply pkg/frontend/show_account.go]: 已更新

File: pkg/frontend/show_account.go (hunk 7 above)
Comment (gouhongshen) [reply pkg/frontend/show_account.go]: 已更新

File: pkg/frontend/show_account.go (hunk 8 above)
Comment (gouhongshen) [reply pkg/frontend/show_account.go]: 已更新

**Review by aptend — APPROVED** [review]

**Review by triump2020 — APPROVED** [review]

**Review by jiangxinmeng1 — APPROVED** [review]

**Review by XuPeng-SH — APPROVED** [review]

**Review by daviszhen — APPROVED** [review]

**Pull request merged**